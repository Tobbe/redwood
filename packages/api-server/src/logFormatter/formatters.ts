import chalk from 'chalk'
import prettyBytes from 'pretty-bytes'
import prettyMs from 'pretty-ms'

const newline = '\n'

const emojiLog: Record<string, string> = {
  warn: '🚦',
  info: '🌲',
  error: '🚨',
  debug: '🐛',
  fatal: '💀',
  trace: '🧵',
}

const isObjectWithData = (object?: Record<string, unknown>) => {
  return object && Object.keys(object).length > 0
}

const isWideEmoji = (character: string) => {
  return character !== '🚦'
}

export const formatBundleSize = (bundleSize: string) => {
  const bytes = parseInt(bundleSize, 10)
  const size = prettyBytes(bytes).replace(/ /, '')
  return chalk.gray(size)
}

export const formatMetadata = (metadata?: Record<string, unknown>) => {
  return isObjectWithData(metadata)
    ? chalk.white(
        newline + '🗒  Metadata' + newline + JSON.stringify(metadata, null, 2)
      )
    : undefined
}

export const formatData = (data?: Record<string, unknown>) => {
  return isObjectWithData(data)
    ? chalk.white(
        newline + '📦 Result Data' + newline + JSON.stringify(data, null, 2)
      )
    : undefined
}

export const formatDate = (instant: number) => {
  const date = new Date(instant)
  const hours = date.getHours().toString().padStart(2, '0')
  const minutes = date.getMinutes().toString().padStart(2, '0')
  const seconds = date.getSeconds().toString().padStart(2, '0')
  const prettyDate = hours + ':' + minutes + ':' + seconds
  return chalk.gray(prettyDate)
}

export const formatLevel = (level: string) => {
  const emoji = emojiLog[level]
  const padding = isWideEmoji(emoji) ? '' : ' '
  return emoji + padding
}

export const formatLoadTime = (elapsedTime?: string) => {
  if (!elapsedTime) {
    return
  }

  const elapsed = Math.round(parseFloat(elapsedTime))
  const time = prettyMs(elapsed)

  return chalk.gray(time)
}

export const formatMessage = (logData: { msg?: string; level: string }) => {
  if (!logData.msg) {
    return
  }

  const msg = formatMessageName(logData.msg)

  switch (logData.level) {
    case 'error':
      return chalk.red(msg)
    case 'trace':
      return chalk.white(msg)
    case 'warn':
      return chalk.magenta(msg)
    case 'debug':
      return chalk.yellow(msg)
    case 'info':
    case 'customLevel':
      return chalk.green(msg)
    case 'fatal':
      return chalk.white.bgRed(msg)
  }

  return msg
}

export const formatMethod = (method: string) => {
  return chalk.white(method)
}

export const formatRequestId = (requestId?: string) => {
  return requestId && chalk.cyan(requestId)
}

export const formatNs = (name?: string) => {
  return name && chalk.cyan(name || '')
}

export const formatName = (name: string) => {
  return chalk.blue(name)
}

export const formatMessageName = (msg: string) => {
  if (msg === 'request') {
    return '<--'
  } else if (msg === 'response') {
    return '-->'
  }

  return msg
}

export const formatOperationName = (operationName?: string) => {
  return operationName && chalk.white(newline + '🏷  ' + operationName)
}

export const formatQuery = (query?: Record<string, unknown>) => {
  return isObjectWithData(query)
    ? chalk.white(
        newline + '🔭 Query' + newline + JSON.stringify(query, null, 2)
      )
    : undefined
}

export const formatResponseCache = (
  responseCache?: Record<string, unknown>
) => {
  return isObjectWithData(responseCache)
    ? chalk.white(
        newline +
          '💾 Response Cache' +
          newline +
          JSON.stringify(responseCache, null, 2)
      )
    : undefined
}

export const formatStatusCode = (statusCode?: string) => {
  statusCode = statusCode || 'xxx'
  return chalk.white(statusCode)
}

export const formatTracing = (data?: Record<string, unknown>) => {
  return isObjectWithData(data)
    ? chalk.white(
        newline + '⏰ Timing' + newline + JSON.stringify(data, null, 2)
      )
    : undefined
}

export const formatUrl = (url?: string) => {
  return url && chalk.white(url)
}

export const formatUserAgent = (userAgent?: string) => {
  return userAgent && chalk.grey(newline + '🕵️‍♀️ ' + userAgent)
}
