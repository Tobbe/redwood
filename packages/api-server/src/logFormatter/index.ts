import jsonParse from 'fast-json-parse'

import { DEFAULT_API_SERVER_LOGGER_NAME } from '../app'

import {
  formatOperationName,
  formatQuery,
  formatData,
  formatResponseCache,
  formatLoadTime,
  formatDate,
  formatNs,
  formatName,
  formatRequestId,
  formatMessage,
  formatTracing,
  formatUserAgent,
  formatMethod,
  formatUrl,
  formatStatusCode,
  formatMetadata,
  formatLevel,
  formatBundleSize,
} from './formatters'

const newline = '\n'

const isPinoLog = (log?: unknown) => {
  return log && Object.prototype.hasOwnProperty.call(log, 'level')
}

export const LogFormatter = () => {
  /**
   * Example inputData:
   *   {
   *     "level": 20,
   *     "time": 1650298442951,
   *     "pid": 80980,
   *     "hostname": "Tobbes-MacBook-Pro.local",
   *     "foo": "foo",
   *     "nested": {
   *       "arr": [
   *         "one",
   *         "two",
   *         "three"
   *       ]
   *     },
   *     "msg": "Example debug log"
   *   }
   *
   *   {
   *     "level": 20,
   *     "time": 1651913678887,
   *     "pid": 83106,
   *     "hostname": "Tobbes-MacBook-Pro.local",
   *     "name": "rw-graphql-server",
   *     "operationName": "BlogPostsQuery",
   *     "query": {},
   *     "requestId": "req-4",
   *     "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.147 Safari/537.36",
   *     "data": {
   *       "blogPosts": [
   *         {
   *           "id": 1,
   *           "title": "A little more about me",
   *           "body": "Raclette shoreditch ...",
   *           "createdAt": "2022-05-07T07:41:21.150Z",
   *           "__typename": "Post"
   *         },
   *         {
   *           "id": 2,
   *           "title": "What is the meaning of life?",
   *           "body": "Meh waistcoat succulents ...",
   *           "createdAt": "2022-05-07T07:41:21.150Z",
   *           "__typename": "Post"
   *         },
   *         {
   *           "id": 3,
   *           "title": "Welcome to the blog!",
   *           "body": "I'm baby single- origin ...",
   *           "createdAt": "2022-05-07T07:41:21.150Z",
   *           "__typename": "Post"
   *         }
   *       ]
   *     },
   *     "responseCache": {
   *       "hit": false,
   *       "didCache": true,
   *       "ttl": null
   *     },
   *     "msg": "GraphQL execution completed: BlogPostsQuery"
   *   }
   */
  const parse = (inputData: string | Record<string, unknown>) => {
    let logData

    if (typeof inputData === 'string') {
      const parsedData = jsonParse(inputData)

      // `value` and `err` here comes from jsonParse above
      if (!parsedData.value || parsedData.err || !isPinoLog(parsedData.value)) {
        return inputData + newline
      }

      logData = parsedData.value
    } else if (isPinoLog(inputData)) {
      // We only end up in here in our tests
      logData = inputData
    } else {
      return inputData + newline
    }

    if (!logData.level) {
      return inputData + newline
    }

    if (typeof logData.level === 'number') {
      logData.level = convertLogNumber(logData)
    }

    return output(logData) + newline
  }

  const convertLogNumber = (logData: { level: number }) => {
    const levelMap: Record<number, string> = {
      10: 'trace',
      20: 'debug',
      30: 'info',
      40: 'warn',
      50: 'error',
      60: 'fatal',
    }

    return levelMap[logData.level] || 'customLevel'
  }

  const noEmpty = (value: string | undefined) => {
    return !!value
  }

  interface GraphqlServerLogs {
    level: string
    name: string
    msg: string
    operationName?: string
    query?: Record<string, unknown>
    requestId?: string
    userAgent?: string
    data?: Record<string, unknown>
    responseCache?: Record<string, unknown>
    tracing?: Record<string, unknown>
  }

  function isGraphqlServerLogs(
    logData: Record<string, unknown> | GraphqlServerLogs
  ): logData is GraphqlServerLogs {
    return (
      (logData as unknown as GraphqlServerLogs).name === 'rw-graphql-server'
    )
  }

  /**
   * Example logData
   * {
   *   level: 'debug',
   *   time: 1651911200534,
   *   pid: 81356,
   *   hostname: 'Tobbes-MacBook-Pro.local',
   *   name: 'rw-graphql-server',
   *   msg: 'GraphQL execution started: BlogPostsQuery',
   * }
   *
   * {
   *   level: 'debug',
   *   time: 1651911200547,
   *   pid: 81356,
   *   hostname: 'Tobbes-MacBook-Pro.local',
   *   name: 'rw-graphql-server',
   *   msg: 'GraphQL execution completed: BlogPostsQuery',
   * }
   *
   * With more logging options enabled in graphql.ts
   * {
   *   level: 'debug',
   *   time: 1651911945301,
   *   pid: 82077,
   *   hostname: 'Tobbes-MacBook-Pro.local',
   *   name: 'rw-graphql-server',
   *   operationName: 'BlogPostsQuery',
   *   query: {},
   *   requestId: 'req-8',
   *   userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.147 Safari/537.36',
   *   data: { blogPosts: [ [Object], [Object], [Object] ] },
   *   responseCache: { hit: false, didCache: true, ttl: null },
   *   msg: 'GraphQL execution completed: BlogPostsQuery',
   * }
   */
  const outputGraphqlServer = (logData: GraphqlServerLogs) => {
    const output: (string | undefined)[] = []

    output.push(formatName(logData.name))
    output.push(formatRequestId(logData.requestId))
    output.push(formatMessage(logData))
    output.push(formatOperationName(logData.operationName))
    output.push(formatQuery(logData.query))
    output.push(formatData(logData.data))
    output.push(formatResponseCache(logData.responseCache))
    output.push(formatTracing(logData.tracing))
    output.push(formatUserAgent(logData.userAgent))

    return output
  }

  interface ApiServerLogs {
    level: string
    name: string
    req?: Record<string, string>
    res?: Record<string, string>
    statusCode?: string
    responseTime?: string
    elapsed?: string
    method?: string
    contentLength?: string
    url?: string
    msg: string
  }

  function isApiServerLogs(
    logData: Record<string, unknown> | ApiServerLogs
  ): logData is ApiServerLogs {
    return (
      (logData as unknown as ApiServerLogs).name ===
      DEFAULT_API_SERVER_LOGGER_NAME
    )
  }

  /**
   * Example logData
   *
   * {
   *   level: 'info',
   *   time: 1651910010935,
   *   pid: 80419,
   *   hostname: 'Tobbes-MacBook-Pro.local',
   *   name: 'api-server',
   *   reqId: 'req-7',
   *   res: { statusCode: 200 },
   *   responseTime: 4.725124835968018,
   *   msg: 'request completed',
   * }
   *
   * {
   *   level: 'info',
   *   time: 1651910010930,
   *   pid: 80419,
   *   hostname: 'Tobbes-MacBook-Pro.local',
   *   name: 'api-server',
   *   reqId: 'req-7',
   *   req: {
   *     method: 'POST',
   *     url: '/graphql',
   *     hostname: 'localhost:8910',
   *     remoteAddress: '::1',
   *     remotePort: 64980
   *   },
   *   msg: 'incoming request',
   * }
   */
  const outputApiServer = (logData: ApiServerLogs) => {
    const output: (string | undefined)[] = []

    output.push(formatName(logData.name))

    if (logData.req) {
      output.push(formatMethod(logData.req.method))
      output.push(formatUrl(logData.req.url))
    }

    if (logData.res) {
      output.push(formatStatusCode(logData.res.statusCode))
      output.push(formatLoadTime(logData.responseTime))
    }

    output.push(formatMessage(logData))

    return output
  }

  const outputMetadata = (logData: Record<string, unknown>) => {
    const {
      time: _time,
      level: _level,
      pid: _pid,
      hostname: _hostname,
      msg: _msg,
      ns: _ns,
      ...metadata
    } = logData

    return [formatMetadata(metadata)]
  }

  interface LogData {
    time: number
    msg?: string
    level?: string
    elapsed?: string
    contentLength?: string
    ns?: string
    err?: { stack: string }
    stack?: string
  }

  const output = (logData: LogData & GraphqlServerLogs & ApiServerLogs) => {
    let output: (string | undefined)[] = []

    if (!logData.level) {
      logData.level = 'customLevel'
    }

    output.push(formatDate(logData.time || Date.now()))
    output.push(formatLevel(logData.level))

    if (isGraphqlServerLogs(logData)) {
      output = output.concat(outputGraphqlServer(logData))
    } else if (isApiServerLogs(logData)) {
      output = output.concat(outputApiServer(logData))
    }

    // TODO: When is this used?
    const elapsed = logData.elapsed
    if (elapsed) {
      output.push(formatLoadTime(elapsed))
    }

    // TODO: When is this used?
    const contentLength = logData.contentLength
    if (contentLength) {
      output.push(formatBundleSize(contentLength))
    }

    // TODO: When is this used?
    output.push(formatNs(logData.ns))

    if (!isGraphqlServerLogs(logData) && !isApiServerLogs(logData)) {
      output.push(formatMessage(logData))
      // Metadata can for example come from user logs, like
      // logger.debug({ user }, `User ${user.id}`)
      output = output.concat(outputMetadata(logData))
    }

    return output.filter(noEmpty).join(' ')
  }

  return parse
}
