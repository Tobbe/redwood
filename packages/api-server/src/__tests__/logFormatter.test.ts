import chalk from 'chalk'

import { LogFormatter } from '../logFormatter/index'

const logFormatter = LogFormatter()

describe('LogFormatter', () => {
  describe('Formats log levels as emoji', () => {
    test('Formats Trace level', () => {
      expect(logFormatter({ level: 10 })).toMatch('üßµ')
    })

    test('Formats Debug level', () => {
      expect(logFormatter({ level: 20 })).toMatch('üêõ')
    })

    test('Formats Info level', () => {
      expect(logFormatter({ level: 30 })).toMatch('üå≤')
    })

    test('Formats Warn level', () => {
      expect(logFormatter({ level: 40 })).toMatch('üö¶')
    })

    test('Formats Error level', () => {
      expect(logFormatter({ level: 50 })).toMatch('üö®')
    })
  })

  describe('Formats log messages', () => {
    test('Formats newline-delimited json data with a message', () => {
      expect(
        logFormatter({ level: 10, message: 'Message in a bottle' })
      ).toMatch('Message in a bottle')
    })

    test('Formats newline-delimited json data with a msg', () => {
      expect(logFormatter({ level: 10, msg: 'Message in a bottle' })).toMatch(
        'Message in a bottle'
      )
    })

    test('Formats a text message', () => {
      expect(logFormatter('Handles text data')).toMatch('Handles text data')
    })

    test('Formats Get Method and Status Code', () => {
      const logData = { level: 10, method: 'GET', statusCode: 200 }
      expect(logFormatter(logData)).toMatch('GET')
      expect(logFormatter(logData)).toMatch('200')
    })

    test('Formats Post Method and Status Code', () => {
      const logData = { level: 10, method: 'POST', statusCode: 200 }
      expect(logFormatter(logData)).toMatch('POST')
      expect(logFormatter(logData)).toMatch('200')
    })
  })

  describe('Formats GraphQL injected log data from useRedwoodLogger plugin', () => {
    test('Handles query', () => {
      expect(
        logFormatter({
          level: 10,
          query: {
            id: 1,
          },
        })
      ).toMatch('"id": 1')
    })

    test('Handles operation name', () => {
      expect(
        logFormatter({ level: 10, operationName: 'GET_BLOG_POST_BY_ID' })
      ).toMatch('GET_BLOG_POST_BY_ID')
    })

    test('Handles GraphQL data', () => {
      expect(
        logFormatter({
          level: 10,
          data: { post: { id: 1, title: 'My Blog Post' } },
        })
      ).toMatch('My Blog Post')
    })

    test('Handles browser user agent', () => {
      const userAgent =
        'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.1 Safari/605.1.15'
      expect(
        logFormatter({
          level: 10,
          userAgent,
        })
      ).toMatch(/Mozilla.*AppleWebKit.*Safari/)
    })
  })

  describe('Custom log data', () => {
    test('Should include the custom log attribute text', () => {
      expect(
        logFormatter({
          level: 10,
          custom: 'I should see this custom message text',
        })
      ).toMatch('I should see this')
    })

    test('Should include the custom log attribute info with a special emoji and label', () => {
      expect(
        logFormatter({
          level: 10,
          custom: 'I should see this custom emoji and label',
        })
      ).toMatch('üóí  Metadata')
    })

    test('Should include the custom log attribute info with nested text message', () => {
      expect(
        logFormatter({
          level: 10,
          custom: {
            msg: 'I should see this custom message in the log',
          },
        })
      ).toMatch('I should see this custom message in the log')
    })
  })

  test('Should include the custom log attribute info with a number attribute', () => {
    expect(
      logFormatter({
        level: 10,
        custom: {
          msg: 'I should see this custom message and number in the log',
          number: 100,
        },
      })
    ).toMatch('100')
  })

  test('Should include the custom log attribute info with a nested object attribute', () => {
    expect(
      logFormatter({
        level: 10,
        custom: {
          msg: 'I should see this custom object in the log',
          obj: { foo: 'bar' },
        },
      })
    ).toMatch('"foo": "bar"')
  })

  test('Should include the custom log attribute info with a nested object attribute', () => {
    expect(
      logFormatter({
        level: 10,
        custom: {
          msg: 'I should see this custom object in the log',
          obj: { foo: 'bar' },
        },
      })
    ).toMatch('"foo": "bar"')
  })

  describe('text input', () => {
    it('Should format graphql logs with data, requestId, userAgent and operationName enabled, plus response cache configured', () => {
      expect(
        logFormatter(
          '{"level":20,"time":1652375352844,"pid":41826,"hostname":"Tobbes-MacBook-Pro.local","name":"rw-graphql-server","operationName":"BlogPostsQuery","query":{},"requestId":"req-4","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.147 Safari/537.36","data":{"blogPosts":[{"id":1,"title":"A little more about me","body":"Raclette shoreditch before they sold out lyft. Ethical bicycle rights meh prism twee. Tote bag ennui vice, slow-carb taiyaki crucifix whatever you probably haven\'t heard of them jianbing raw denim DIY hot chicken. Chillwave blog succulents freegan synth af ramps poutine wayfarers yr seitan roof party squid. Jianbing flexitarian gentrify hexagon portland single-origin coffee raclette gluten-free. Coloring book cloud bread street art kitsch lumbersexual af distillery ethical ugh thundercats roof party poke chillwave.","createdAt":"2022-05-07T07:41:21.150Z","__typename":"Post"},{"id":2,"title":"What is the meaning of life?","body":"Meh waistcoat succulents umami asymmetrical, hoodie post-ironic paleo chillwave tote bag. Trust fund kitsch waistcoat vape, cray offal gochujang food truck cloud bread enamel pin forage. Roof party chambray ugh occupy fam stumptown. Dreamcatcher tousled snackwave, typewriter lyft unicorn pabst portland blue bottle locavore squid PBR&B tattooed.","createdAt":"2022-05-07T07:41:21.150Z","__typename":"Post"},{"id":3,"title":"Welcome to the blog!","body":"I\'m baby single- origin coffee kickstarter lo - fi paleo skateboard.Tumblr hashtag austin whatever DIY plaid knausgaard fanny pack messenger bag blog next level woke.Ethical bitters fixie freegan,helvetica pitchfork 90\'s tbh chillwave mustache godard subway tile ramps art party. Hammock sustainable twee yr bushwick disrupt unicorn, before they sold out direct trade chicharrones etsy polaroid hoodie. Gentrify offal hoodie fingerstache.","createdAt":"2022-05-07T07:41:21.150Z","__typename":"Post"}]},"responseCache":{"hit":false,"didCache":true,"ttl":null},"msg":"GraphQL execution completed: BlogPostsQuery"}'
        )
      ).toEqual(
        [
          `${chalk.gray('19:09:12')} üêõ ${chalk.blue(
            'rw-graphql-server'
          )} ${chalk.cyan('req-4')} ${chalk.yellow(
            'GraphQL execution completed: BlogPostsQuery'
          )} `,
          chalk.white('\nüè∑  BlogPostsQuery') + ' ',
          chalk.white(
            [
              '\nüì¶ Result Data',
              '{',
              '  "blogPosts": [',
              '    {',
              '      "id": 1,',
              '      "title": "A little more about me",',
              '      "body": "Raclette shoreditch before they sold out lyft. Ethical bicycle rights meh prism twee. Tote bag ennui vice, slow-carb taiyaki crucifix whatever you probably haven\'t heard of them jianbing raw denim DIY hot chicken. Chillwave blog succulents freegan synth af ramps poutine wayfarers yr seitan roof party squid. Jianbing flexitarian gentrify hexagon portland single-origin coffee raclette gluten-free. Coloring book cloud bread street art kitsch lumbersexual af distillery ethical ugh thundercats roof party poke chillwave.",',
              '      "createdAt": "2022-05-07T07:41:21.150Z",',
              '      "__typename": "Post"',
              '    },',
              '    {',
              '      "id": 2,',
              '      "title": "What is the meaning of life?",',
              '      "body": "Meh waistcoat succulents umami asymmetrical, hoodie post-ironic paleo chillwave tote bag. Trust fund kitsch waistcoat vape, cray offal gochujang food truck cloud bread enamel pin forage. Roof party chambray ugh occupy fam stumptown. Dreamcatcher tousled snackwave, typewriter lyft unicorn pabst portland blue bottle locavore squid PBR&B tattooed.",',
              '      "createdAt": "2022-05-07T07:41:21.150Z",',
              '      "__typename": "Post"',
              '    },',
              '    {',
              '      "id": 3,',
              '      "title": "Welcome to the blog!",',
              '      "body": "I\'m baby single- origin coffee kickstarter lo - fi paleo skateboard.Tumblr hashtag austin whatever DIY plaid knausgaard fanny pack messenger bag blog next level woke.Ethical bitters fixie freegan,helvetica pitchfork 90\'s tbh chillwave mustache godard subway tile ramps art party. Hammock sustainable twee yr bushwick disrupt unicorn, before they sold out direct trade chicharrones etsy polaroid hoodie. Gentrify offal hoodie fingerstache.",',
              '      "createdAt": "2022-05-07T07:41:21.150Z",',
              '      "__typename": "Post"',
              '    }',
              '  ]',
              '}',
            ].join('\n')
          ) + ' ',
          chalk.white(
            [
              '\nüíæ Response Cache',
              '{',
              '  "hit": false,',
              '  "didCache": true,',
              '  "ttl": null',
              '}',
            ].join('\n')
          ) + ' ',
          chalk.gray(
            '\nüïµÔ∏è‚Äç‚ôÄÔ∏è Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.147 Safari/537.36'
          ),
          '\n',
        ].join('')
      )
    })

    it('Should format basic graphql logs', () => {
      expect(
        logFormatter(
          '{"level":20,"time":1652380326869,"pid":46072,"hostname":"Tobbes-MacBook-Pro.local","name":"rw-graphql-server","msg":"GraphQL execution completed: BlogPostsQuery"}'
        )
      ).toMatchSnapshot()
    })

    it('Should format api server request logs', () => {
      expect(
        logFormatter(
          '{"level":30,"time":1652380326855,"pid":46072,"hostname":"Tobbes-MacBook-Pro.local","name":"api-server","reqId":"req-5","req":{"method":"POST","url":"/graphql","hostname":"localhost:8910","remoteAddress":"::1","remotePort":64671},"msg":"incoming request"}'
        )
      ).toMatchSnapshot()
    })

    it('Should format api server response logs', () => {
      expect(
        logFormatter(
          '{"level":30,"time":1652380326869,"pid":46072,"hostname":"Tobbes-MacBook-Pro.local","name":"api-server","reqId":"req-4","res":{"statusCode":200},"responseTime":48.36454105377197,"msg":"request completed"}'
        )
      ).toMatchSnapshot()
    })

    it('Should format regular user logs that has metadata', () => {
      // This is the result of a Redwood app calling
      // logger.debug({ input }, 'updatePost input')
      expect(
        logFormatter(
          '{"level":20,"time":1652381509868,"pid":46484,"hostname":"Tobbes-MacBook-Pro.local","input":{"title":"A little more about me","body":"Raclette shoreditch before they sold out lyft."},"msg":"updatePost input"}'
        )
      ).toMatchSnapshot()
    })

    it('Should print only metadata, if only metadata is logged', () => {
      expect(
        logFormatter(
          '{"level":20,"time":1652387012644,"pid":49648,"hostname":"Tobbes-MacBook-Pro.local","input":{"title":"A little more about me","body":"Metadata only"}}'
        )
      ).toEqual(
        [
          `${chalk.gray('22:23:32')} üêõ` + ' ',
          chalk.white(
            [
              '\nüóí  Metadata',
              '{',
              '  "input": {',
              '    "title": "A little more about me",',
              '    "body": "Metadata only"',
              '  }',
              '}',
            ].join('\n')
          ),
          '\n',
        ].join('')
      )
    })

    it('Should print the full details of error objects', () => {
      expect(
        logFormatter(
          '{"level":50,"time":1652385852132,"pid":48711,"hostname":"Tobbes-MacBook-Pro.local","err":{"code":3,"message":"Error number three"},"msg":"We got an error"}'
        )
      ).toEqual(
        [
          `${chalk.gray('22:04:12')} üö® ${chalk.red('We got an error')}` + ' ',
          chalk.white(
            [
              '\nüóí  Metadata',
              '{',
              '  "err": {',
              '    "code": 3,',
              '    "message": "Error number three"',
              '  }',
              '}',
            ].join('\n')
          ),
          '\n',
        ].join('')
      )
    })

    it('Should not crash on malformed json', () => {
      expect(
        logFormatter(
          '{"level":50,"time":1652385852132,"msg":"missing ending quote and curly brace'
        )
      ).toEqual(
        '{"level":50,"time":1652385852132,"msg":"missing ending quote and curly brace\n'
      )
    })
  })
})
