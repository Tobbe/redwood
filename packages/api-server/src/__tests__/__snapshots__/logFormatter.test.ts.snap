// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogFormatter text input Should format api server request logs 1`] = `
"[90m20:32:06[39m 🌲 [32mincoming request[39m [37m[39m
[37m🗒  Metadata[39m
[37m{[39m
[37m  \\"name\\": \\"api-server\\",[39m
[37m  \\"reqId\\": \\"req-5\\",[39m
[37m  \\"req\\": {[39m
[37m    \\"method\\": \\"POST\\",[39m
[37m    \\"url\\": \\"/graphql\\",[39m
[37m    \\"hostname\\": \\"localhost:8910\\",[39m
[37m    \\"remoteAddress\\": \\"::1\\",[39m
[37m    \\"remotePort\\": 64671[39m
[37m  }[39m
[37m}[39m
"
`;

exports[`LogFormatter text input Should format api server response logs 1`] = `
"[90m20:32:06[39m 🌲 [32mrequest completed[39m [37m[39m
[37m🗒  Metadata[39m
[37m{[39m
[37m  \\"name\\": \\"api-server\\",[39m
[37m  \\"reqId\\": \\"req-4\\",[39m
[37m  \\"res\\": {[39m
[37m    \\"statusCode\\": 200[39m
[37m  },[39m
[37m  \\"responseTime\\": 48.36454105377197[39m
[37m}[39m
"
`;

exports[`LogFormatter text input Should format basic graphql logs 1`] = `
"[90m20:32:06[39m 🐛 [34mrw-graphql-server[39m [33mGraphQL execution completed: BlogPostsQuery[39m
"
`;

exports[`LogFormatter text input Should format regular user logs that has metadata 1`] = `
"[90m20:51:49[39m 🐛 [33mupdatePost input[39m [37m[39m
[37m🗒  Metadata[39m
[37m{[39m
[37m  \\"input\\": {[39m
[37m    \\"title\\": \\"A little more about me\\",[39m
[37m    \\"body\\": \\"Raclette shoreditch before they sold out lyft.\\"[39m
[37m  }[39m
[37m}[39m
"
`;
